source/particle/interpolator/bilinear_least_squares.cc

```diff
@@ -89,14 +90,17 @@ namespace aspect
 
         // Noticed that the size of matrix A is n_particles x matrix_dimension
         // which usually is not a square matrix. Therefore, we find the
-        // least squares solution of Ax=r by solving the "normal" equations
-        // (A^TA) x = A^Tr.
+        // least squares solution of Ax=r by solving reduced QR factorization
+        // Ax = QRc = b -> Q^TQRc = Rc =Q^Tb
+        dealii::ImplicitQR<dealii::Vector<double>> qr;
         const unsigned int matrix_dimension = (dim == 2) ? 4 : 8;
```

Comment:
Not your fault, but now that I think about it `matrix_dimension` is not really a helpful variable name for a matrix that is not a square matrix. Is there a better name? What about `n_matrix_columns`?

Reply:
Yes, we know how many columns A will have.

In one dimension A has two columns;, namely the number of unknowns we are solving for.  Also, the first column is all 1's, but that might not be relevant here.  (However, it is relevant to a proof that A can't be rank deficient in 1 dimension, which of course we don't use, but it may be possible to extend this fact to two and three dimensions to prove that A can't be rank deficient for bilinear LS.  
However, it's going to take me some time to follow up on this. I gave my first online (Zoom) lecture of the Spring Quarter to 90 students yesterday and it was a disaster. :-(  Feel free to work on this problem if you like.)

In two dimensions the number of columns in A is again the number of unknowns, say q.  This integer q will depend on whether we are keeping the xy term in bilinear LS or not. ( I don't think I need to write the bilinear approximation out for @gassmoeller or @sac-bsa.)

Again, same in 3D.  I'm kind of rushed, so I'll think about a good variable name, but perhaps one of you will come up with one first?

Also note that A^TA will be a q x q matrix.  (Why doesn't GitHub have a LaTeX editor?)

Reply:
I think it would be easier to read if you used `compositional_field.get_fe().dofs_per_cell` or whatever the correct call is. That way, it's clear that the dimension of the matrix equals the number of unknowns you're trying to determine. That *happens* to be equal to 4 and 8, but makes it *clear* what it is semantically.